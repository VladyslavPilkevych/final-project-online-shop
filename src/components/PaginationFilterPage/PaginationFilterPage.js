import React, { memo, useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import styles from './PaginationFilterPage.module.scss';
import CardItem from '../CardItem/CardItem';
import {
  setFilterPaginationPage,
} from '../../store/actionCreators/filterAC';

function PaginationFilterPage(props) {
  const dispatch = useDispatch();
  const { filterProducts, itemsPerPage } = props;
  const filterPaginationPage = useSelector((state) => state.filter.filterPaginationPage);
  const [currentItems, setCurrentItems] = useState(null);
  const [pageCount, setPageCount] = useState(0);
  const [itemOffset, setItemOffset] = useState(0);

  useEffect(() => {
    if (filterPaginationPage === 0) {
      setItemOffset(0);
    }
    const endOffset = itemOffset + itemsPerPage;
    if (filterProducts !== null) {
      setCurrentItems(filterProducts.slice(itemOffset, endOffset));
      setPageCount(Math.ceil(filterProducts.length / itemsPerPage));
    }
  }, [itemOffset, itemsPerPage, filterProducts, filterPaginationPage]);
  const handlePageClick = (event) => {
    window.scrollTo(0, 0);
    const newOffset = (event.selected * itemsPerPage) % filterProducts.length;
    setItemOffset(newOffset);
    dispatch(setFilterPaginationPage(event.selected));
  };
  return (
    <>
      <div className={styles.filterItem}>
        {currentItems && currentItems.map((product) => (
          <div key={product.itemNo} className={styles.gridItemContainer}>
            <CardItem
              name={product.name}
              currentPrice={product.currentPrice}
              id={product._id}
              img={product.imageUrls[0]}
              quantity={product.quantity}
              previousPrice={product.previousPrice}
              model={product.model}
              itemNo={product.itemNo}
            />
          </div>
        ))}
      </div>
      {currentItems?.length === 0 && <h1 className={styles.noItems}>No items</h1>}
      <ReactPaginate
        forcePage={filterPaginationPage}
        breakLabel="..."
        nextLabel="›"
        previousLabel="‹"
        onPageChange={handlePageClick}
        pageCount={pageCount}
        renderOnZeroPageCount={null}
        containerClassName={styles.paginateContainer}
        activeLinkClassName={styles.activeLinkClassName}
      />
    </>
  );
}

PaginationFilterPage.propTypes = {
  filterProducts: PropTypes.array,
  itemsPerPage: PropTypes.number,
};

PaginationFilterPage.defaultProps = {
  filterProducts: [],
  itemsPerPage: 0,
};

export default memo(PaginationFilterPage);