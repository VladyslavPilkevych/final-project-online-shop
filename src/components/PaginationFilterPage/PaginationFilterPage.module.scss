@import "../../assets/styles/global.module.scss";

@mixin fonts($font-family, $color, $bold: 700, $size: 16px) {
    font-family: $font-family, sans-serif;
    font-weight: $bold;
    font-size: $size;
    color: $color;
}

.filterItem {
    margin-top: 20px;
    display: grid;
    width: 100%;
    grid-template-columns: repeat(5, 1fr);
}
.gridItemContainer {
    width: 100%;
    height: 100%;
}

.paginateContainer {
    display: flex;
    margin: 55px auto;
    justify-content: center;
    li {
        cursor: pointer;
        margin: 0 5px;
        a {
            @include fonts("Poppins", $dark-grey-text, 600);
            padding: 8px 14px;
            border: 2px solid $dark-grey-text;
            border-radius: 50%;
            width: 15px;
        }
    }
}
.activeLinkClassName {
    color: $black !important;
    border: none !important;
    background-color: $bg-white-color;
}
.noItems {
    @include fonts("Poppins", $black, 600, 26px);
    text-align: center;
    margin-top: 10%;
}

@media (min-width: 768px) and (max-width: 1024px) {
    .filterItem {
        grid-template-columns: repeat(4, 1fr);
    }
}
@media (min-width: 426px) and (max-width: 768px) {
    .filterItem {
        grid-template-columns: repeat(3, 1fr);
    }
    .paginateContainer {
        li {
            margin: 5px;
        }
    }
}
@media (max-width: 426px) {
    .filterItem {
        grid-template-columns: repeat(2, 1fr);
    }
    .paginateContainer {
        li {
            margin: 5px;
        }
    }
}
