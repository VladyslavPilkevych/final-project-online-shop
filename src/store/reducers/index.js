import { combineReducers } from 'redux';
import preloaderReducer from './preloaderReducer';
import productsReducer from './productsReducer';
import menuReducer from './menuReducer';
import userReducer from './userReducer';
import searchReducer from './searchReducer';
import cartReducer from './cartReducer';
import miniMenuReducer from './miniMenuReducer';
import filtersCategoriesReducer from './filtersCategoriesReducer';

const reducer = combineReducers({
  loader: preloaderReducer,
  products: productsReducer,
  menu: menuReducer,
  miniMenu: miniMenuReducer,
  filtersCategories: filtersCategoriesReducer,
  user: userReducer,
  search: searchReducer,
  cart: cartReducer,
});

export default reducer;
