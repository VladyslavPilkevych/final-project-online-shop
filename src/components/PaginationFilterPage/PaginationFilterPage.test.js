import { render } from '@testing-library/react';
import PaginationFilterPage from './PaginationFilterPage';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';

function Component() {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <PaginationFilterPage />
      </Provider>
    </BrowserRouter>
  );
}

describe('Snapshot PaginationFilterPage should render', () => {
  test('Snapshot PaginationFilterPage is rendered', () => {
    const { asFragment } = render(<Component />);
    expect(asFragment()).toMatchSnapshot();
  });
});