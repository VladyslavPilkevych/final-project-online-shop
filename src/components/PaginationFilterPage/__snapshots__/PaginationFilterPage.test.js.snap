// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot PaginationFilterPage should render Snapshot PaginationFilterPage is rendered 1`] = `
<DocumentFragment>
  <div
    class="filterItem"
  />
  <h1
    class="noItems"
  >
    No items
  </h1>
  <ul
    class="paginateContainer"
  >
    <li
      class="previous disabled"
    >
      <a
        aria-disabled="true"
        aria-label="Previous page"
        class=" "
        rel="prev"
        role="button"
        tabindex="-1"
      >
        ‹
      </a>
    </li>
    <li
      class="next"
    >
      <a
        aria-disabled="false"
        aria-label="Next page"
        class=""
        rel="next"
        role="button"
        tabindex="0"
      >
        ›
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
